dubbo-serialization/dubbo-serialization-protobuf/src/main/java/org/apache/dubbo/common/serialize/protobuf/ProtobufObjectInput.java

```diff
@@ -0,0 +1,157 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dubbo.common.serialize.protobuf;
+
+import io.protostuff.ProtobufIOUtil;
+import io.protostuff.Schema;
+import io.protostuff.runtime.RuntimeSchema;
+import org.apache.dubbo.common.serialize.ObjectInput;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Type;
+
+public class ProtobufObjectInput implements ObjectInput {
+
+    private DataInputStream dis;
+
+    public ProtobufObjectInput(InputStream inputStream) {
+        dis = new DataInputStream(inputStream);
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    @Override
+    public Object readObject() throws IOException, ClassNotFoundException {
+        int classNameLength = dis.readInt();
+        int bytesLength = dis.readInt();
+
+        if (classNameLength < 0 || bytesLength < 0) {
+            throw new IOException();
+        }
+
+        byte[] classNameBytes = new byte[classNameLength];
+        dis.readFully(classNameBytes, 0, classNameLength);
+
+        byte[] bytes = new byte[bytesLength];
+        dis.readFully(bytes, 0, bytesLength);
+
+        String className = new String(classNameBytes);
+        Class clazz = Class.forName(className);
+
+        Object result;
+        if (Utils.WRAPPER_SET.contains(clazz) || clazz.isArray() || clazz.isEnum()) {
+            Schema<Wrapper> schema = RuntimeSchema.getSchema(Wrapper.class);
+            Wrapper wrapper = schema.newMessage();
+            ProtobufIOUtil.mergeFrom(bytes, wrapper, schema);
+            result = wrapper.getData();
+        } else {
+            Schema schema = RuntimeSchema.getSchema(clazz);
+            result = schema.newMessage();
+            ProtobufIOUtil.mergeFrom(bytes, result, schema);
+        }
+
+        return result;
+    }
+
+    @Override
+    public <T> T readObject(Class<T> clazz) throws IOException, ClassNotFoundException {
+        int classNameLength = dis.readInt();
+        int bytesLength = dis.readInt();
+
+        if (classNameLength < 0 || bytesLength < 0) {
+            throw new IOException();
+        }
+
+        byte[] classNameBytes = new byte[classNameLength];
+        dis.read(classNameBytes, 0, classNameLength);
+
+        byte[] bytes = new byte[bytesLength];
+        dis.read(bytes, 0, bytesLength);
+
+        T result;
+        if (Utils.WRAPPER_SET.contains(clazz) || clazz.isArray() || clazz.isEnum()) {
+            Schema<Wrapper> schema = RuntimeSchema.getSchema(Wrapper.class);
+            Wrapper wrapper = schema.newMessage();
+            ProtobufIOUtil.mergeFrom(bytes, wrapper, schema);
+            result = (T) wrapper.getData();
+        } else {
+            Schema<T> schema = RuntimeSchema.getSchema(clazz);
+            result = schema.newMessage();
+            ProtobufIOUtil.mergeFrom(bytes, result, schema);
+        }
+
+        return result;
```

Comment:
advice to refactor here by method extraction because here is a lot of replicated codes.